📖 Cluster API State Metrics proposal

chrischdi opened the pull request:
Signed-off-by: Christian Schlotter <schlotterc@vmware.com>

**What this PR does / why we need it**:

Proposal to address parts of #1477. Introduces experimental component `cluster-api-state-metrics` which exposes metrics

chrischdi commented:
I checked the mentioned links and was able to reach them

/retest

killianmuldoon commented:
>This doesn't restart the gh action tests :-(

Yeah - it's a bother - we have an issue here: https://github.com/kubernetes/test-infra/issues/25831

These can be restarted manually and will re-run on another push. The MD link checker is flaky so it's probably fine, so I don't think there's anything to worry about here :smile:

enxebre commented on docs/proposals/20220411-cluster-api-state-metrics.md:
```diff
@@ -0,0 +1,323 @@
+| `capi_machinedeployment_spec_replicas`                               | `.spec.replicas`                              | Gauge |                                                     | [Deployment]                      |
+| `capi_machinedeployment_spec_strategy_rollingupdate_max_unavailable` | `.spec.strategy.rollingUpdate.maxUnavailable` | Gauge |                                                     | [Deployment]                      |
+| `capi_machinedeployment_spec_strategy_rollingupdate_max_surge`       | `.spec.strategy.rollingUpdate.maxSurge`       | Gauge |                                                     | [Deployment]                      |
+| `capi_machinedeployment_status_phase`                                | `.status.phase==<phase>`                      | Gauge | `phase=<phase>`                                     | [Pod], [machinedeployment phases] |
```
I think we'll want to track some 1 particular conditions here instead

chrischdi replied:
I also added the condition metric for the other resources. Looks like I also missed adding the initial metrics for MHC too..

Thanks for the review 👍

killianmuldoon reviewed (approved):
/lgtm

Can't wait to get this into the repo!

fabriziopandini commented:
/lgtm 
Looking forward to getting this under /exp and in our tilt setup.
As this is accomplished I think we should consider
- to document how to install this new component, so we can start getting feedback from the broader audience
- providing some pre-defined Prometheus dashboard
But can come as a follow up iteration

chrischdi commented:
I identified and fixed some small things:

* added a `_paused` metric, this is already implemented in https://github.com/mercedes-benz/cluster-api-state-metrics
* fixed formatting for some tables
* fixed version reference for the `capi_machine_info` metric

chrischdi commented:
Also added already implemented `capi_machineset_status_available_replicas` metric

chrischdi commented:
I squashed the commits :-)

enxebre commented on docs/proposals/20220411-cluster-api-state-metrics.md:
```diff
@@ -0,0 +1,359 @@
+  - 70% of my Machines are blocked on deletion.
+
+## Proposal
+
```
All this section does a great job explaining what will be scrapped and why but can we get into the impl details of how cluster-api-state-metrics will work? diagrams are useful as well.

chrischdi replied:
Kindly see other comment and additional commit :-)

enxebre commented on docs/proposals/20220411-cluster-api-state-metrics.md:
```diff
@@ -0,0 +1,359 @@
+
+#### Reusage of kube-state-metrics packages
+
+The proposed exporter should re-use as many packages provided by kube-state-metrics as possible. This allows to re-use flags, configuration and extended functionality like sharding without additional implementation.
```
Is this stating will have a vendored dep with kube-state-metrics?

What is an exported? how does kube-state-metrics metrics work?

Can we elaborate the details https://github.com/kubernetes-sigs/cluster-api/pull/6404/files#r849729360?

chrischdi replied:
Yes, we will re-use packages and functions from kube-state-metrics to not re-implement it :-)

l:169 scratches the surface of implementation details. I will iterate and improve this to go into more details. 👍

chrischdi replied:
I added more detailed paragraphs. Also replaced the usage of the word `exporter` to reduce confusion :-) .

I don't know which kind of diagram would be helpful to visualize unclear details.
A full existing implementation is already in the referenced, to be contributed project: https://github.com/mercedes-benz/cluster-api-state-metrics

I would be happy if you could take another look :-)

chrischdi commented:
The test `Check Markdown links / Broken Links (pull_request)` looks again like a false positive :-(

enxebre commented:
thanks lgtm after squashing, we can always iterate later as we need it.

chrischdi commented:
Squashed the commits :-)

fabriziopandini commented:
/lgtm

enxebre commented:
/lgtm

chrischdi commented:
@apricote / @johannesfrey / @tobiasgiese / @sbueringer : kindly asking you to take another look if there are any more objections or if we should proceed :-)

Otherwise I would propose a lazy consensus until Wednesday next week (27th April).

tobiasgiese commented:
/lgtm

I already reviewed your changes and I‘m totally fine with them 👍🏻 Thanks for taking care!

schrej commented:
We've been using the source project for a while and are very happy with it, so
/lgtm

And for that lazy consensus
/hold

johannesfrey commented:
Yep, looks really good. Also fine with the lazy consensus. Thank you very much!

/lgtm

apricote commented on docs/proposals/20220411-cluster-api-state-metrics.md:
```diff
@@ -0,0 +1,389 @@
+
+#### Package structure for cluster-api-state-metrics
+
+- `/exp/pkg/store` contains the metric implementation exposed by the `customresource.RegistryFactory`[[2]] interface.
```
Is this path supposed to be relative to the project root? If yes, we might have a problem with go module boundaries.

In case we have source code in `/exp/pkg/store` and `/exp/state-metrics` the `go.mod` for these should be in `/exp`, but there is already some code in `/exp` that belong to the root `/go.mod` (`MachinePool`, `ClusterResourceSet`).

I suggest to move `/exp/pkg` to `/exp/state-metrics/pkg`.

chrischdi replied:
good finding, no it was supposed to land under `exp/state-metrics/pkg/store` :+1:

chrischdi commented:
> We've been using the source project for a while and are very happy with it, so /lgtm
> 
> And for that lazy consensus /hold

I'd prefer to all get this time to take a look if they want to ;-)

/hold

apricote commented:
/lgtm

Thanks a lot!

fabriziopandini commented:
/lgtm

seanschneeweiss commented on docs/proposals/20220411-cluster-api-state-metrics.md:
```diff
@@ -0,0 +1,389 @@
+
+This component is derived from  [mercedes-benz/cluster-api-state-metrics] and all the merits for its inception goes to the team that created it: [@chrischdi](https://github.com/chrischdi), [@seanschneeweiss](https://github.com/seanschneeweiss), [@tobiasgiese](https://github.com/tobiasgiese).
+
+A special thank goes to [Mercedes Benz] who allowed the donation of this project to CNCF and to the Cluster API community.
```
```suggestion
A special thank goes to [Mercedes-Benz] who allowed the donation of this project to CNCF and to the Cluster API community.
```

seanschneeweiss commented on docs/proposals/20220411-cluster-api-state-metrics.md:
```diff
@@ -0,0 +1,389 @@
+|----------------------------------------------------------------------|-----------------------------------------------|-------|-----------------------------------------------------|-----------------------------------|
+| `capi_machinedeployment_labels`                                      | `1`                                           | Gauge | `.metadata.labels`                                  | common                            |
+| `capi_machinedeployment_created`                                     | `.metadata.creationTimestamp`                 | Gauge |                                                     | common                            |
+| `capi_machinedeployment_paused`                                      | `annotations.HasPaused()`                     | Gauge |                                                     |                                   |
```
MachineDeployments also support pausing by setting `spec.paused`.

seanschneeweiss reviewed (commented):
Great work - so happy to see this being integrated into CAPI.

chrischdi commented:
> Great work - so happy to see this being integrated into CAPI.

Thanks for these (hopefully) last findings :-)

chrischdi commented:
Broken Links check has again false-positives:
```
FILE: ./docs/book/src/clusterctl/provider-contract.md
[✖] https://help.github.com/en/github/administering-a-repository/creating-releases

7 links checked.

ERROR: 1 dead links found!
[✖] https://help.github.com/en/github/administering-a-repository/creating-releases → Status: 403
```

Github Action issue: https://github.com/gaurav-nelson/github-action-markdown-link-check/issues/136

sbueringer commented:
Great work folks!

/lgtm

@fabriziopandini WDYT can we merge this PR given that the deadline is today? (https://github.com/kubernetes-sigs/cluster-api/pull/6404#issuecomment-1104237433)

johannesfrey commented:
Thanks again!

/lgtm

tobiasgiese commented:
/retest
/lgtm

Thanks @chrischdi 🙂

seanschneeweiss commented:
/lgtm

Thank you @chrischdi

fabriziopandini commented:
/approve
/lgtm
🎉

sbueringer commented:
/hold cancel

sbueringer commented:
Looks like the linkchecker consistently fails since 2 days: https://github.com/kubernetes-sigs/cluster-api/actions/workflows/lint-docs.yaml
(the action is not run on every PR)
